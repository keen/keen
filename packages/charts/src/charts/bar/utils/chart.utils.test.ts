import {
  register as registerTimezone,
  unregister as unregisterTimezone,
} from 'timezone-mock';

import {
  generateHorizontalGroupedBars,
  generateVerticalGroupedBars,
  Options,
} from './chart.utils';

import { verticalBarChart, horizontalBarChart } from '../bar-chart.fixtures';

const data = [
  { label: 'January', sale: -3, buy: 11, revenue: 30 },
  { label: 'February', sale: 12, buy: 3, revenue: 21 },
];

beforeAll(() => {
  registerTimezone('UTC');
});

afterAll(() => {
  unregisterTimezone();
});

describe('generateHorizontalGroupedBars()', () => {
  const chart: Options = {
    data,
    ...horizontalBarChart,
  };

  test('should create proper domain for xScale', () => {
    const { xScale } = generateHorizontalGroupedBars(chart);

    expect(xScale.domain()).toMatchInlineSnapshot(`
      Array [
        -4,
        12,
      ]
    `);
  });

  test('should increase domain for xScale', () => {
    const data = [
      { label: 'Cats', adopted: 12 },
      { label: 'Dogs', adopted: 17 },
      { label: 'Horses', adopted: 2 },
    ];

    const { xScale } = generateHorizontalGroupedBars({
      data,
      ...verticalBarChart,
      keys: ['adopted'],
    });

    expect(xScale.domain()).toMatchInlineSnapshot(`
      Array [
        0,
        18,
      ]
    `);
  });

  test('should create proper domain for yScale', () => {
    const { yScale } = generateHorizontalGroupedBars(chart);

    expect(yScale.domain()).toMatchInlineSnapshot(`
      Array [
        "February",
        "January",
      ]
    `);
  });

  test('should not create bars for disabled keys', () => {
    const data = [
      { label: 'Marketing', people: 10, rooms: 3 },
      { label: 'Customer Success', people: 16, rooms: 10 },
    ];

    const { bars } = generateHorizontalGroupedBars({
      data,
      ...horizontalBarChart,
      keys: ['people', 'rooms'],
      disabledKeys: ['rooms'],
    });

    const result = [
      { key: '0.people', selector: [0, 'people'] },
      { key: '1.people', selector: [1, 'people'] },
    ];

    expect(bars).toMatchObject(result);
  });

  test('should sort bars in "ascending" order', () => {
    const data = [
      { label: 'Marketing', people: 10, rooms: 3 },
      { label: 'Customer Success', people: 2, rooms: 10 },
    ];

    const { bars } = generateHorizontalGroupedBars({
      data,
      ...horizontalBarChart,
      keys: ['people', 'rooms'],
      barsOrder: 'ascending',
    });

    expect(bars).toMatchSnapshot();
  });

  test('should sort bars in "descending" order', () => {
    const data = [
      { label: 'Marketing', people: 15, rooms: 34 },
      { label: 'Customer Success', people: 21, rooms: 10 },
    ];

    const { bars } = generateHorizontalGroupedBars({
      data,
      ...horizontalBarChart,
      keys: ['people', 'rooms'],
      barsOrder: 'descending',
    });

    expect(bars).toMatchSnapshot();
  });

  test('should generate horizontal grouped bars with one missing color', () => {
    const data = [
      { label: 'Marketing', people: 15, rooms: 34 },
      { label: 'Customer Success', people: 21, rooms: 10 },
    ];

    const { bars } = generateHorizontalGroupedBars({
      data,
      ...horizontalBarChart,
      keys: ['people', 'rooms'],
      barsOrder: 'descending',
      colors: ['red'],
    });

    expect(bars).toMatchSnapshot();
  });
});

describe('generateVerticalGroupedBars()', () => {
  const chart: Options = {
    data,
    ...verticalBarChart,
  };

  test('should normalize intervals dates', () => {
    const data = [
      { label: '2020-01-01T12:00:00.000Z', value: 30 },
      { label: '2020-01-01T13:00:00.000Z', value: 21 },
    ];

    const { xScale } = generateVerticalGroupedBars({
      ...chart,
      keys: ['value'],
      labelSelector: 'label',
      xScaleSettings: {
        type: 'band',
        precision: 'hour',
        useUTC: false,
      },
      data,
    });

    expect(xScale.domain()).toMatchInlineSnapshot(`
      Array [
        "2020-01-01T12:00:00.000Z",
        "2020-01-01T13:00:00.000Z",
      ]
    `);
  });

  test('should create proper domain for xScale', () => {
    const { xScale } = generateVerticalGroupedBars(chart);

    expect(xScale.domain()).toMatchInlineSnapshot(`
      Array [
        "January",
        "February",
      ]
    `);
  });

  test('should create proper domain for yScale', () => {
    const { yScale } = generateVerticalGroupedBars(chart);

    expect(yScale.domain()).toMatchInlineSnapshot(`
      Array [
        -5,
        30,
      ]
    `);
  });

  test('should not create bars for disabled keys', () => {
    const data = [
      { label: 'Marketing', people: 10, rooms: 3, cars: 10 },
      { label: 'Customer Success', people: 16, rooms: 10, cars: 12 },
    ];

    const { bars } = generateHorizontalGroupedBars({
      data,
      ...verticalBarChart,
      keys: ['people', 'rooms', 'cars'],
      disabledKeys: ['rooms'],
    });

    const result = [
      { key: '0.people', selector: [0, 'people'] },
      { key: '0.cars', selector: [0, 'cars'] },
      { key: '1.people', selector: [1, 'people'] },
      { key: '1.cars', selector: [1, 'cars'] },
    ];

    expect(bars).toMatchObject(result);
  });

  test('should increase domain for yScale', () => {
    const data = [
      { label: 'January', revenue: 33 },
      { label: 'February', revenue: 21 },
      { label: 'March', revenue: 25 },
    ];

    const { yScale } = generateVerticalGroupedBars({
      data,
      ...verticalBarChart,
      keys: ['revenue'],
    });

    expect(yScale.domain()).toMatchInlineSnapshot(`
      Array [
        0,
        35,
      ]
    `);
  });

  test('should sort bars in "ascending" order', () => {
    const data = [
      { label: 'Marketing', people: 10, rooms: 3 },
      { label: 'Customer Success', people: 2, rooms: 10 },
    ];

    const { bars } = generateVerticalGroupedBars({
      data,
      ...verticalBarChart,
      keys: ['people', 'rooms'],
      barsOrder: 'ascending',
    });

    expect(bars).toMatchSnapshot();
  });

  test('should sort bars in "descending" order', () => {
    const data = [
      { label: 'Marketing', people: 15, rooms: 34 },
      { label: 'Customer Success', people: 21, rooms: 10 },
    ];

    const { bars } = generateVerticalGroupedBars({
      data,
      ...verticalBarChart,
      keys: ['people', 'rooms'],
      barsOrder: 'descending',
    });

    expect(bars).toMatchSnapshot();
  });

  test('should generate vertical grouped bars with one missing color', () => {
    const data = [
      { label: 'Marketing', people: 15, rooms: 34 },
      { label: 'Customer Success', people: 21, rooms: 10 },
    ];

    const { bars } = generateVerticalGroupedBars({
      data,
      ...verticalBarChart,
      keys: ['people', 'rooms'],
      barsOrder: 'descending',
      colors: ['red'],
    });

    expect(bars).toMatchSnapshot();
  });
});
