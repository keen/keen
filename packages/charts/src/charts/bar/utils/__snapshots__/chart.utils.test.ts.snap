// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateHorizontalGroupedBars() should sort bars in "ascending" order 1`] = `
Array [
  Object {
    "color": "red",
    "height": 66.25,
    "key": "0.people",
    "selector": Array [
      0,
      "people",
    ],
    "value": 10,
    "width": 380,
    "x": 10,
    "y": 81.25,
  },
  Object {
    "color": "green",
    "height": 66.25,
    "key": "0.rooms",
    "selector": Array [
      0,
      "rooms",
    ],
    "value": 3,
    "width": 114,
    "x": 10,
    "y": 15,
  },
  Object {
    "color": "red",
    "height": 66.25,
    "key": "1.people",
    "selector": Array [
      1,
      "people",
    ],
    "value": 2,
    "width": 76,
    "x": 10,
    "y": 147.5,
  },
  Object {
    "color": "green",
    "height": 66.25,
    "key": "1.rooms",
    "selector": Array [
      1,
      "rooms",
    ],
    "value": 10,
    "width": 380,
    "x": 10,
    "y": 213.75,
  },
]
`;

exports[`generateHorizontalGroupedBars() should sort bars in "descending" order 1`] = `
Array [
  Object {
    "color": "red",
    "height": 66.25,
    "key": "0.people",
    "selector": Array [
      0,
      "people",
    ],
    "value": 15,
    "width": 162.85714285714286,
    "x": 10,
    "y": 81.25,
  },
  Object {
    "color": "green",
    "height": 66.25,
    "key": "0.rooms",
    "selector": Array [
      0,
      "rooms",
    ],
    "value": 34,
    "width": 369.1428571428571,
    "x": 10,
    "y": 15,
  },
  Object {
    "color": "red",
    "height": 66.25,
    "key": "1.people",
    "selector": Array [
      1,
      "people",
    ],
    "value": 21,
    "width": 228,
    "x": 10,
    "y": 147.5,
  },
  Object {
    "color": "green",
    "height": 66.25,
    "key": "1.rooms",
    "selector": Array [
      1,
      "rooms",
    ],
    "value": 10,
    "width": 108.57142857142856,
    "x": 10,
    "y": 213.75,
  },
]
`;

exports[`generateVerticalGroupedBars() should sort bars in "ascending" order 1`] = `
Array [
  Object {
    "color": "red",
    "height": 370,
    "key": "0.people",
    "selector": Array [
      0,
      "people",
    ],
    "value": 10,
    "width": 120,
    "x": 130,
    "y": 10,
  },
  Object {
    "color": "blue",
    "height": 111,
    "key": "0.rooms",
    "selector": Array [
      0,
      "rooms",
    ],
    "value": 3,
    "width": 120,
    "x": 10,
    "y": 269,
  },
  Object {
    "color": "red",
    "height": 74,
    "key": "1.people",
    "selector": Array [
      1,
      "people",
    ],
    "value": 2,
    "width": 120,
    "x": 250,
    "y": 306,
  },
  Object {
    "color": "blue",
    "height": 370,
    "key": "1.rooms",
    "selector": Array [
      1,
      "rooms",
    ],
    "value": 10,
    "width": 120,
    "x": 370,
    "y": 10,
  },
]
`;

exports[`generateVerticalGroupedBars() should sort bars in "descending" order 1`] = `
Array [
  Object {
    "color": "red",
    "height": 158.57142857142858,
    "key": "0.people",
    "selector": Array [
      0,
      "people",
    ],
    "value": 15,
    "width": 120,
    "x": 130,
    "y": 221.42857142857142,
  },
  Object {
    "color": "blue",
    "height": 359.42857142857144,
    "key": "0.rooms",
    "selector": Array [
      0,
      "rooms",
    ],
    "value": 34,
    "width": 120,
    "x": 10,
    "y": 20.571428571428577,
  },
  Object {
    "color": "red",
    "height": 222,
    "key": "1.people",
    "selector": Array [
      1,
      "people",
    ],
    "value": 21,
    "width": 120,
    "x": 250,
    "y": 158,
  },
  Object {
    "color": "blue",
    "height": 105.71428571428572,
    "key": "1.rooms",
    "selector": Array [
      1,
      "rooms",
    ],
    "value": 10,
    "width": 120,
    "x": 370,
    "y": 274.2857142857143,
  },
]
`;
